polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -771,6 +759,60 @@ Value LoopInternalization::getTileSize(LoopLikeOpInterface loop) const {
                                                 tileSize);
 }
 
+void LoopInternalization::transform(FunctionOpInterface func,
+                                    const FunctionKernelInfo &funcKernelInfo) {
+  auto gpuModule = func->getParentOfType<gpu::GPUModuleOp>();
+
+  // Ensure there is local memory to be used.
+  unsigned localMemoryRemain = getLocalMemoryRemain(gpuModule, localMemorySize);
+  if (localMemoryRemain == 0)
+    return;
+
+  // Calculate the required local memory for all accesses in
+  // loopToSharedMemref to be promoted.
+  SmallVector<gpu::GPUFuncOp> kernels;
+  funcKernelInfo.getKernelCallers(func, kernels);
+  sycl::ReqdWorkGroupSize reqdWorkGroupSize(kernels);
+  Optional<unsigned> reqdLocalMemory =
+      getReqdLocalMemory(loopToSharedMemref, reqdWorkGroupSize);
+  if (reqdLocalMemory.has_value() && *reqdLocalMemory > localMemoryRemain) {
+    LLVM_DEBUG(llvm::dbgs() << "Not enough local memory\n");
+    return;
+  }
+
+  memref::GlobalOp workGroupLocalMemory = getWorkGroupLocalMemory(
+      gpuModule,
+      reqdLocalMemory.has_value() ? *reqdLocalMemory : localMemoryRemain);
+
+  // Get or create work group size.
+  Value ndItem = getNdItemArgument(func);
+  OpBuilder builder(func->getRegion(0));
+  WorkGroupSize workGroupSize(ndItem, reqdWorkGroupSize, builder);
+
+  // Create local id.
+  auto ndItemTy = cast<sycl::NdItemType>(
+      cast<MemRefType>(ndItem.getType()).getElementType());
```

Comment:
Did you concider `SYCLLocalIDOp` (https://github.com/intel/llvm/blob/sycl-mlir/mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td#L182) ?

I'm fine with what I see as the rounded range kernel thingy is making things a bit more complex. But could be more efficient to not have to pull data from somewhere.

Reply:
done.